Checks selection

nelsonkopliku opened the pull request:
# Description

Please include a summary of the changes and the related issue, if it exists.
Also, include relevant motivation and context for this PR. List any dependencies that are required for this change.

Fixes # (issue)

## Did you add the right label?

Remember to add the right labels to this PR.

- [ ] **DONE**

## How was this tested?

Describe the tests that have been added/changed for this new behavior.

- [ ] **DONE**

## Did you update the documentation?

Remember to ask yourself if your PR requires changes to the following documentation:

- [Manual installation guide](https://github.com/trento-project/docs/blob/main/guides/manual-installation.md)
- [Trento Ansible guide](https://github.com/trento-project/ansible/blob/main/README.md)
- [Trento Wanda guides](https://github.com/trento-project/web/tree/main/guides)

Add a documentation PR or write that no changes are required for the documentation.

- [ ] **DONE**

EMaksy commented on lib/wanda/checks_selection.ex:
```diff
@@ -0,0 +1,78 @@
+        custom_values
+
+      _ ->
+        nil
```
Should this not be a list  as map_values/2 expects  a list and in worst case it receives a nil

nelsonkopliku replied:
`nil` here means _the customization was not found_.

When not found values falls back to an empty list as defined in the second argument of `Enum.find_value` on line 32
```elixir
Enum.find_value(available_customizations, [], fn ...
```

EMaksy commented on lib/wanda/checks_selection.ex:
```diff
@@ -0,0 +1,78 @@
+    env
+    |> Catalog.get_catalog()
+    |> Enum.map(
+      &%SelectableCheck{
```
This is a bit harder to read,  what about extracting the check building  to a function ?

nelsonkopliku replied:
Done 2d9ef9d8b980b22d50ba3bf3c298caf373453ea7

EMaksy reviewed (commented):
Hey great work, just left some comments about Wanda.ChecksSelection

arbulu89 reviewed (commented):
Hey @nelsonkopliku 
Just wondering if we need a new context and controller files.
Doesn't make sense to add this logic in the `catalog.ex` code or same folder?
At the end, we are simply "upgrading" the `get_catalog` function to use a `group_id` and load the customized values. At the end we are still getting the catalog with some enriched data. We could have `get_catalog_for_group` or something.
I don't know... I wouldn't add new contexts and files just to have files with less functions.

We could even to enhance the current catalog endpoint to send the values, and the difference with the group specific one can only have the customized values.

Just thinking out loud..

PD: I got a bit confused with the `current_value` and `customization` terms

nelsonkopliku commented on lib/wanda/catalog/selectable_check.ex:
```diff
@@ -0,0 +1,39 @@
+  @type customized_value :: %{
+          name: String.t(),
+          customizable: true,
+          current_value: boolean() | number() | String.t(),
```
`current_value` is going to be the resolved value given the context. It could be the default or one of the conditional values.

I like `customized_value`, gonna change to that, thanks.

arbulu89 replied:
But is this happening now? Or is yet to be implemented?
Because as far as I understand the code you are putting the default value into `current_value` if the value is customizable.

nelsonkopliku commented on lib/wanda/catalog/selectable_check.ex:
```diff
@@ -0,0 +1,39 @@
+  defstruct [
+    :id,
+    :name,
+    :group,
```
ah no sorry, this is not the `group_id`, it's actually the group as in `Corosync`

nelsonkopliku commented on lib/wanda/checks_selection.ex:
```diff
@@ -0,0 +1,85 @@
+         } = value,
+         custom_values
+       ) do
+    case Enum.find(custom_values, &(&1.name == value_name)) do
```
here we go https://github.com/trento-project/wanda/pull/563/commits/f81aa661befb352056298d25d4acd633809b0576

nelsonkopliku commented on lib/wanda_web/schemas/v1/checks_selection/selectable_checks_response.ex:
```diff
@@ -0,0 +1,56 @@
+            properties: %{
+              id: %Schema{type: :string, description: "Check ID"},
+              name: %Schema{type: :string, description: "Check name"},
+              group: %Schema{type: :string, description: "Check group"},
```
As above, that's the actual check group.

nelsonkopliku commented on lib/wanda_web/controllers/v1/checks_selection_controller.ex:
```diff
@@ -0,0 +1,45 @@
+      ]
+    ],
+    responses: [
+      ok: {"Selectable checks response", "application/json", SelectableChecksResponse}
```
Not really, if the group does not exist in the customization table it means there are no customization available, not that there are not selectable checks for the given env.

arbulu89 replied:
Ok, i understand the hassle, but it would be strange to return empty lists for group_ids that we don't even have.
One thing is not having check for that group and other not having the group itself

nelsonkopliku replied:
I see, however for wanda the group id is transient, nothing really matching anything in its state.
We don't have anything to say hey, we don't know anything about this group id, so here a 404 for you.

In this case the group id is necessary to load customizations for a set of checks loaded by env.

nelsonkopliku commented:
Hey @arbulu89 thanks for the review.
I addressed feedback here https://github.com/trento-project/wanda/pull/563/commits/a99118d32d5c8c0315b12c5a255969d1bfd64fdb and here https://github.com/trento-project/wanda/pull/563/commits/f81aa661befb352056298d25d4acd633809b0576

I could move the functions in catalog context quite easily, however with controllers is more cumbersome with versioning in place, so just simpler having a separated one.

nelsonkopliku commented:
@arbulu89 I made anyway an attempt to move everything into the same catalog controller here 5f7d0e4e1e73b896ef6675d9c157ecb578971cad.
I can live with it, if needed we can move things around again later.
